Add note about runtime disabling Memfault (#106)
memfault/memfault-docs: docs/mcu/runtime-control.mdx

---
id: runtime-control
title: Enabling and Disabling Memfault at Runtime
sidebar_label: Runtime Control
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

This guide covers how to enable and disable Memfault at runtime (vs. at
compile-time). This can vary widely depending on the particular integration
strategy or system, so some examples are provided after the general information
section.

## Overall Strategy

> The simplest way to deactivate Memfault at runtime is to turn off the mechanism
> used to send Memfault chunks to the cloud. This means disabling wherever in the
> system the packetizer APIs are used (see
> [here](/docs/mcu/data-from-firmware-to-the-cloud) for details).

Reviewer: Maybe:
```suggestion
The recommended way to deactivate Memfault at runtime is to turn off the mechanism used to send Memfault chunks to the cloud. 

This allows you to pull Memfault diagnostic data off of devices manually if you
have a direct connection to the device such as JTAG, serial, etc or re-enable dynamically and get historical context immediately.

To accomplish this one needs to add a check around the system the packetizer APIs are used (see [here](/docs/mcu/data-from-firmware-to-the-cloud) for details).
```

For example, you might use a variable to control whether sending is attempted
(perhaps loaded from NVRAM settings database on device):

```c
#include <stdbool.h>

> static bool enable_memfault_upload = false;

Reviewer: ```suggestion
static bool s_enable_memfault_upload = false;
```

...
void periodic_task(void) {

>   if (enable_memfault_upload) {

Reviewer: ```suggestion
  if (s_enable_memfault_upload) {
```

    try_send_memfault_data();
  }
}
```

> Note that this only disables the sending of Memfault data. All the collection
> mechanisms will still be enabled as usual (eg coredump saving, metrics, etc).
> 
> This allows you to pull Memfault diagnostic data off of devices manually if you
> have a direct connection to the device such as JTAG, serial, etc.
> 

Reviewer: maybe: I moved this up above
```suggestion
```

> handlers), and outside of the scope of this how-to guide.

Reviewer 2: I'd re-frame this as a benefit instead of saying it's hard. 

> This allows you to pull Memfault diagnostic data off of devices manually if you have a direct connection to the device such as JTAG, serial, etc. 

Or something along those lines.

Author: good call!

## Zephyr (including nRF Connect SDK)

This will depend on if `CONFIG_MEMFAULT_HTTP_PERIODIC_UPLOAD=y` is used.

### CONFIG_MEMFAULT_HTTP_PERIODIC_UPLOAD=y

> These instructions are based on the example in the nrf-connect SDK
> (https://github.com/nrfconnect/sdk-nrf/tree/master/samples/nrf9160/memfault):

Reviewer: ```suggestion
These instructions are based on [the example in the nrf-connect SDK](https://github.com/nrfconnect/sdk-nrf/tree/master/samples/nrf9160/memfault):
```
